# *Application of Bootstrap samples in Random Forest*

<b>There will be some functions that start with the word "grader" ex: grader_samples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

### Importing relevant packages

In [1]:
!pip install numpy==1.21.2

In [2]:
import warnings
warnings.filterwarnings("ignore")

import numpy as np # import numpy for numerical computation
from sklearn.datasets import load_boston # Using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # import mean_squared_error metric

In [3]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [4]:
x.shape

(506, 13)

In [5]:
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

## *Task 1*

### Step - 1

##### Creating samples
- Randomly create 30 samples from the whole boston data points.

- Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points.

- For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]

##### Create 30 samples 

- Note that as a part of the Bagging when you are taking the random samples make sure each of the sample will have different set of columns<br>

- Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...

- Make sure each sample will have atleast 3 feautres/columns/attributes

### Step - 2 

#### Building High Variance Models on each of the sample and finding train MSE value

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

### Step - 3 

#### Calculating the OOB score

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

## Task 2

#### Computing CI of OOB Score and Train MSE
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intervals of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence interval</li>
</ol>

## Task 3

#### Given a single query point predict the price of house

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

<br><br>

## Task 1

### Step - 1

#### Creating samples - Algorithm

![alt text](https://i.imgur.com/BTVYXQ1.jpg/)

#### Code for generating samples

In [6]:
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 
    
    # select 303 rows randomly without replacement from the current dataset having 506 rows 
    selected_rows = np.random.choice(506, 303, replace = False)
    
    # replicating 203 rows from the 303 randomly selected rows 
    replicated_rows = np.random.choice(selected_rows, 203, replace = False)
    
    # selecting random columns between 3 and 13 indices
    columns = np.random.randint(3, 13, size = 1)
    selected_columns = np.random.choice(range(13), size = columns, replace = False)
    
    sample_data = input_data[selected_rows[:, None], selected_columns]
    
    target_sample_data = target_data[selected_rows]
    
    
    # Replicating Data

    replicated_sample_data = input_data[replicated_rows[:, None], selected_columns]
    
    target_replicated_data = target_data[replicated_rows]
    
    
    # Concatenating Data
    
    final_sample_data = np.vstack((sample_data, replicated_sample_data))
                                  
    final_target_data = np.vstack((target_sample_data.reshape(-1, 1), target_replicated_data.reshape(-1, 1)))
    
    # To calculate the OOB score
    oob_rows = np.array([i for i in range(0,506) if i not in selected_rows])
    oob_data = input_data[oob_rows[:,None], selected_columns]


    return final_sample_data.tolist(), final_target_data.tolist(), selected_rows.tolist(), selected_columns.tolist(), oob_data.tolist()
    # return sampled_input_data, sampled_target_data, selected_rows, selected_columns
    # NOTE: please return as lists

In [7]:
# Print the lengths of the lists generated

a, b, c, d, e = generating_samples(x, y)

print(len(a)) 
print(len(b)) 
print(len(c)) 
print(len(d)) 
print(len(e)) 

506
506
303
4
203


##### Grader function - 1 

In [8]:
def grader_samples(a, b, c, d, e):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length = (len(d)>=3)
    oob_length = (len(e) == 203)
    assert(length and sampled and rows_length and column_length and oob_length)
    return True
a, b, c, d, e = generating_samples(x, y)
grader_samples(a, b, c, d, e)

True

#### Create 30 samples 

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [9]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list

def get_sample_lists(x, y):
    list_input_data = []
    list_output_data = []
    list_selected_row = []
    list_selected_columns = []
    list_oob_data = []

    for i in range(0, 30):
        a, b, c, d, e = generating_samples(x, y)
        list_input_data.append(a)
        list_output_data.append(b)
        list_selected_row.append(c)
        list_selected_columns.append(d)
        list_oob_data.append(e)
    return list_input_data, list_output_data, list_selected_row, list_selected_columns, list_oob_data

In [10]:
list_input_data, list_output_data, list_selected_row, list_selected_columns, get_oob_data = get_sample_lists(x, y)

##### Grader function - 2 

In [11]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

### Step - 2 

<font color='black'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

#### Code for building regression trees

In [12]:
from sklearn.tree import DecisionTreeRegressor # import DecisionTreeRegressor

def model_train_gen(list_input_data, list_output_data):
    list_models_decision_trees = []
    for i in range(0, 30):
        clf =  DecisionTreeRegressor(max_depth=None)
        clf.fit(list_input_data[i], list_output_data[i])
        list_models_decision_trees.append(clf)
    return list_models_decision_trees

<font color='black'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

#### Code for calculating MSE

In [13]:
def get_MSE(model_list, list_selected_columns, x, y):
    prediction_list=[]
    for i in range(len(list_models_decision_trees)):
        predict = list_models_decision_trees[i].predict(x[:,list_selected_columns[i]])
        prediction_list.append(predict.tolist())
    predict_array_np = np.array(prediction_list)
    final_predict_np = (np.sum(predict_array_np,axis=0))/(len(prediction_list))
    MSE = np.sum(np.power((y - final_predict_np ),2))/(len(final_predict_np))
    return MSE

In [14]:
def mse_score_gen(x,models,selected_columns):
    y_pred_for_mse = []

    for i in x:
        y_pred = []

        for j in range(len(models)):
            y_pred.append(models[j].predict(i[selected_columns[j]].reshape(1,-1)))
        y_pred_for_mse.append(np.median(np.asarray(y_pred)))
    
    mse = mean_squared_error(y,np.asarray(y_pred_for_mse))
    return mse

In [15]:
selected_models = model_train_gen(list_input_data, list_output_data)

In [16]:
mse = mse_score_gen(x, selected_models, list_selected_columns)
print(mse)

0.036897233201580996


### Step - 3

<font color='black'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

####  Code for calculating OOB score

In [17]:
def oob_score_gen(x,models,selected_columns,selected_rows):
    y_pred_for_oob=[]

    for i in range(len(x)):
        a=x[i]
        y_pred =[]

        for j in range(len(models)):
            if i not in selected_rows[j]:
                y_pred.append(models[j].predict(a[selected_columns[j]].reshape(1,-1)))
        y_pred_for_oob.append(np.median(np.asarray(y_pred)))
    oob = mean_squared_error(y,np.asarray(y_pred_for_oob))
    return oob

In [18]:
oob = oob_score_gen(x, selected_models, list_selected_columns, list_selected_row)
print(oob)

14.282852162933683


<br>

## Task 2

#### Code for computing Confidence Intervals

In [19]:
MSE_arr = []
OOB_arr = []
list_input_data =[]
list_output_data =[]
list_selected_rows =[]
list_selected_columns =[]
list_oob_data =[]

for j in range(35):

    for i in range(30):
        o,p,q,r,s = generating_samples(x,y)
        list_input_data.append(o)
        list_output_data.append(p)
        list_selected_rows.append(q)
        list_selected_columns.append(r)
        list_oob_data.append(s)
  
    updated_models = model_train_gen(list_input_data, list_output_data)

    mse_val = mse_score_gen(x, updated_models, list_selected_columns)
    MSE_arr.append(mse_val)
  
    oob_val = oob_score_gen(x, updated_models, list_selected_columns, list_selected_rows)
    OOB_arr.append(oob_val)

In [20]:
MSE_arr

[0.057979249011857706,
 0.08346343873517789,
 0.06938241106719373,
 0.08600790513833993,
 0.03737154150197629,
 0.08232707509881423,
 0.05445158102766799,
 0.03904150197628457,
 0.06087944664031622,
 0.05942687747035573,
 0.06389822134387352,
 0.06092885375494072,
 0.0454199604743083,
 0.04787549407114625,
 0.0651828063241107,
 0.054752964426877486,
 0.06138833992094863,
 0.07222826086956519,
 0.06520256916996048,
 0.06112154150197628,
 0.06367094861660083,
 0.04166996047430831,
 0.046902173913043486,
 0.07552865612648223,
 0.049817193675889306,
 0.05971343873517784,
 0.07645750988142293,
 0.0651877470355731,
 0.052020750988142336,
 0.05731225296442687,
 0.0652124505928854,
 0.054970355731225294,
 0.07559288537549405,
 0.04800889328063243,
 0.05686264822134386]

In [21]:
OOB_arr

[13.91105050555165,
 12.842478905738844,
 12.972098876375862,
 12.523880543035725,
 13.378944785895513,
 13.235716292962225,
 13.681451937575355,
 11.824330300117412,
 13.163098366652324,
 12.060955704867846,
 13.656523288310975,
 14.013053351281226,
 12.071046736217543,
 11.901761697245034,
 14.344616190200872,
 13.22501296975531,
 11.855351243244334,
 13.03049514089856,
 13.054875850618833,
 13.285145791992687,
 13.03341167968065,
 13.497954051101924,
 12.996145510533195,
 13.180294562304498,
 13.951789502952446,
 13.35256658859475,
 13.083254357264973,
 13.37202495717351,
 13.16640594641374,
 13.160595669733896,
 13.146612574498393,
 12.923966470129153,
 12.749411410130588,
 11.743523639650453,
 13.913413297094458]

In [22]:
mse_mean = np.mean(MSE_arr)
mse_std_dev = np.std(MSE_arr)
oob_std_dev = np.std(OOB_arr)
oob_mean = np.mean(OOB_arr)
# print(mse_mean)
# print(mse_std_dev)
# print(oob_std_dev)
# print(oob_mean)
MSE_conf_interval_ub = mse_mean+2*(mse_std_dev/np.sqrt(35))
MSE_conf_interval_lb = mse_mean-2*(mse_std_dev/np.sqrt(35))
OOB_conf_interval_ub = oob_mean+2*(oob_std_dev/np.sqrt(35))
OOB_conf_interval_lb = oob_mean-2*(oob_std_dev/np.sqrt(35))
print("MSE Confidence Interval:", MSE_conf_interval_lb,MSE_conf_interval_ub)
print("OOB Confidence Interval:", OOB_conf_interval_lb,OOB_conf_interval_ub)

MSE Confidence Interval: 0.05641792246332673 0.06456824354457842
OOB Confidence Interval: 12.84687427561041 13.284740507006433


<br>

## Task 3

<font color='black'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in Task - 1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

#### Code for Task 3 

In [23]:
xq = [0.18, 20.0, 5.00, 0.0, 0.421, 5.60, 72.2, 7.95, 7.0, 30.0, 19.1, 372.13, 18.60]
xq = np.array(xq).reshape(1,-1)
xq.shape

(1, 13)

In [24]:
selected_models = model_train_gen(list_input_data, list_output_data)

In [25]:
def pred_price(xq, models, selected_columns):
    y_pred_price=[]

    for i in xq:
        y_pred = []

        for j in range(len(models)):
            y_pred.append(models[j].predict(i[selected_columns[j]].reshape(1,-1)))
        y_pred_price.append(np.median(np.asarray(y_pred)))
 
    return y_pred_price

house_price = pred_price(xq, selected_models, list_selected_columns)
print(house_price)

[18.9]


### OBSERVATIONS

##### When trained using 30 samples MSE is 0.036897233201580996 and OOB is 14.282852162933683

##### 95% confidence interval of 35 different samples of MSE is [0.05641792246332673 0.06456824354457842] and OOB is [12.84687427561041 13.284740507006433].

##### Both MSE and OOB of Task 1 are not in range of 95% CI obtained in Task 2.This could be for reasons that samples taken in Task 1 are just 30. But in Task 2 we did the experiment 35 times by taking sample sizes of 30. According to the Cental Limit Theorem, this should reflect population MSE and OOB more accurately.

##### The prediction obtained for the house price given data point is 18.9